File: compiler/bench/shellsort.c
// expected: 49995000
// Shell sort benchmark - sort 10000 elements, sum result

#include <stdio.h>

#define SIZE 10000

int arr[SIZE];

void shell_sort(void) {
    for (int gap = SIZE / 2; gap > 0; gap /= 2) {
        for (int i = gap; i < SIZE; i++) {
            int key = arr[i];
            int j = i;
            while (j >= gap && arr[j - gap] < key) {
                arr[j] = arr[j - gap];
                j -= gap;
            }
            arr[j] = key;
        }
    }
}

void init_arr(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = SIZE - i - 1;
    }
}

long sum_arr(void) {
    long sum = 0;
    for (int i = 0; i < SIZE; i++) {
        sum += arr[i];
    }
    return sum;
}

int main(void) {
    init_arr();
    shell_sort();
    printf("%ld\n", sum_arr());
    return 0;
}

File: compiler/bench/shellsort.fs
\ expected: 49995000
\ Shell sort benchmark - sort 10000 elements, sum result

10000 constant SIZE
create arr SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;

: shell-sort ( -- )
  SIZE 2 /               \ gap
  begin dup 0> while
    SIZE over do         \ gap
      i arr@ i           \ gap key j
      begin
        dup 2 pick >= over 2 pick - arr@ 2 pick < and
      while
        dup 2 pick - arr@ over arr!
        over -
      repeat
      arr! drop
    loop
    2 /
  repeat drop ;

: init-arr ( -- )
  SIZE 0 do
    SIZE i - 1- i arr!
  loop ;

: sum-arr ( -- n )
  0 SIZE 0 do i arr@ + loop ;

: main
  init-arr
  shell-sort
  sum-arr . cr ;
